pkg/deploy/registry/deployconfig/strategy.go

```diff
@@ -39,14 +41,52 @@ func (strategy) AllowUnconditionalUpdate() bool {
 
 // PrepareForCreate clears fields that are not allowed to be set by end users on creation.
 func (strategy) PrepareForCreate(obj runtime.Object) {
-	_ = obj.(*api.DeploymentConfig)
-	// TODO: need to ensure status.latestVersion is not set out of order
+	dc := obj.(*api.DeploymentConfig)
+	dc.Generation = 1
+	dc.Status = api.DeploymentConfigStatus{}
 }
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
 func (strategy) PrepareForUpdate(obj, old runtime.Object) {
-	_ = obj.(*api.DeploymentConfig)
-	// TODO: need to ensure status.latestVersion is not set out of order
+	newDc := obj.(*api.DeploymentConfig)
+	oldDc := old.(*api.DeploymentConfig)
+
+	oldVersion := oldDc.Status.LatestVersion
+	newVersion := newDc.Status.LatestVersion
+	newStatus := newDc.Status
+
+	// Check for imagechange controller change
+	isImageControllerChange := deployutil.IsImageChangeControllerChange(*newDc, *oldDc)
+
+	// Persist status
+	newDc.Status = oldDc.Status
+
+	// If this update comes from the imagechange controller, tolerate status detail updates.
+	if isImageControllerChange {
+		newDc.Status.Details = newStatus.Details
+	}
+
+	switch {
+	case newDc.Generation == oldDc.Generation+1:
+		// New client generation bump; we need to inform the deploymentconfig controller
+		// to run a new deployment - do that by bumping latestVersion.
+		newDc.Status.LatestVersion = oldVersion + 1
+		return
+
+	case newVersion == oldVersion+1:
+		// Older client version bump; we also need to bump generation.
+		newDc.Generation = oldDc.Generation + 1
+		newDc.Status.LatestVersion = oldVersion + 1
+		return
+	}
+
+	// Any changes to the spec or labels, increment the generation number, any changes
+	// to the status should reflect the generation number of the corresponding object
+	// (should be handled by the controller).
+	if !reflect.DeepEqual(oldDc.Spec, newDc.Spec) ||
```

Comment:
I prefer to keep this out of the switch since it's a catch-all case for the hook (essentially the default in the switch). Generation should be bumped either as part of a change in the spec (what these deep equals do), or when clients want a redeployment of the same thing (@smarterclayton convinced me at https://github.com/openshift/origin/issues/6337)